contributors/design-proposals/pod-injection-policy.md

```diff
@@ -0,0 +1,563 @@
+information into every pod spec where it is needed.
+
+## Constraints and Assumptions
+
+1.  Future work might require new mechanisms to be made to work with existing
+    controllers such as deployments and replicasets that create pods. Existing
+    controllers that create pods should recreate their pods when a new Pod Injection
+    Policy is added that would effect them.
+
+## Use Cases
+
+1.  As a user, I want to be able to describe a way that pods should be injected
+    with the information to consume a particular service. This would allow me
+    to concisely model the information about how the service
+    should be consumed without altering every consuming pod spec.
+
+### Loose coupling between services and their consumers
+
+Using a Pod Injection Policy allows pod template authors to not have to explicitly
+set information for every pod. This way authors of pod templates consuming a
+specific service do not need to know all the details about that service.
+
+The pods chosen for injection will be based off the policy's `Selector`.
+
+## Proposed Changes
+
+### PodInjectionPolicy API object
+
+This resource is alpha. The policy itself is immutable. The API group will be
+called `podinjectionpolicy` and the version is `v1alpha1`.
+
+```go
+type PodInjectionPolicy struct {
+    unversioned.TypeMeta
+    ObjectMeta
+
+    Spec PodInjectionPolicySpec
+}
+
+type PodInjectionPolicySpec struct {
+    // Selector defines which Pods should be modified, this only targets Pods.
+    Selector      unversioned.LabelSelector
+    // Env defines the collection of EnvVar to inject.
+    Env           []EnvVar `json:omitempty`
+    // EnvFrom defines the collection of EnvFromSource to inject.
+    EnvFrom       []EnvFromSource `json:omitempty`
+    // Volumes defines the collection of Volume to inject.
+    Volumes       []Volume `json:omitempty`
+    // VolumeMounts defines the collection of VolumeMount to inject.
+    VolumeMounts  []VolumeMount `json:omitempty`
+}
+```
+
+#### Validations
+
+In order for the Pod Injection Policy to be valid it must fulfill the
+following constraints:
+
+- The `Selector` field must be defined. This is how we know which pod templates
+  to inject so therefore it is required.
+- The policy must define _at least_ 1 of `Env`, `EnvFrom`, `Volumes`, or
+  `VolumeMounts`.
+- If you define a `Volume`, it has to define a `VolumeMount`.
+
+This resource will be immutable, if you want to change something you can delete
+the old policy and recreate a new one. We can change this to be mutable in the
+future but by disallowing it now, we will not break people in the future.
+
+#### Conflicts
+
+There are a number of edge conditions that might occur at the time of
+injection. These are as follows:
+
+- Merging volume lists with no conflicts: if a pod already has a volume defined
+  exactly as defined in the PodInjectionPolicy. This would replace the other volume.
```

Comment:
"exactly" meaning all properties or just the mount point?  I would think just the mount point would be enough to check and this would override an existing one.